### PR TITLE
MAGN-8245 Upgrade Helix to .net 4.5

### PR DESCRIPTION
**Pull request by ikeough**
### Purpose

This PR does the following.

- Integrates a version of the Helix Toolkit component which is now the standard. Previously, we had used a custom build of the Helix Toolkit. See [MAGN-8245](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-8245), for a list of our contributions to that project for this purpose.
- Fixes [MAGN-8245](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-8245), by integrating the new version of the Helix Toolkit using .net 4.5, which supports software fallback (WARP) for rendering over Remote Desktop.
- Replaces the Helix Toolkit and SharpDX assembly references to the `\extern` folder with NuGet packages representing the same.

### Impact assessment

- [ ] Changes the way 3rd party nodes work (Add ikeough to the reviewer list)
- [ ] Includes migrations (Add ikeough to the reviewer list)
- [ ] Changes a public interface (Add lukechurch to the reviewer list)
- [ ] Affects which components of the system depends on which other components (Add pboyer to the reviewer list)
- [x] Introduces new components (Add lukechurch to the reviewer list)
- [ ] Has a performance impact (Add keyu to the reviewer list)
- [x] Changes the installer (Add sharadkjaiswal to the reviewer list)
  - The `d3d10warp.dll` and `D3DCompiler_44.dll` files must now be bundled with the installer.
- [ ] User facing strings, if any, are extracted into `*.resx` files

#### References

- See [MAGN-8245](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-8245) for a list of PRs 

#### Testing strategy

- Ensure 3D graphics are visible on Remote Desktop.
- Ensure 3D graphics are visible on machines which do not support DirectX11.
- Ensure 3D graphics are available on machines with "switchable" graphics set to the CPU graphics controller (I.e. Intel Integrated Graphics).

### Review staging

- [ ] Architectural Design Review
- [ ] UI Review

### Reviewers

@pboyer 

### FYIs

@jnealb @sharadkjaiswal

### CODE REVIEW

**Comment by pboyer** [comment]
@ikeough The build fails. 

Your code doesn't seem to following the coding standards as it adds several new components with no XML documentation.

https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards

I would like you to break this into smaller PRs. Is that possible?

File: src/DynamoCoreWpf/ViewModels/Watch3D/DynamoEffectsManager.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
+﻿using HelixToolkit.Wpf.SharpDX;
+using SharpDX.Direct3D11;
+using SharpDX.DXGI;
+
+namespace Dynamo.Wpf.ViewModels.Watch3D
+{
+    public class DynamoEffectsManager : DefaultEffectsManager
```
Comment (pboyer) [inline src/DynamoCoreWpf/ViewModels/Watch3D/DynamoEffectsManager.cs]: This file needs careful documentation. Other developers have no idea what an `EffectsManager` is.

File: src/DynamoCoreWpf/ViewModels/Watch3D/AttachedProperties.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
+﻿using System.Windows;
+
+using HelixToolkit.Wpf.SharpDX;
+
+namespace Dynamo.Wpf.ViewModels.Watch3D
+{
+    public static class AttachedProperties
```
Comment (pboyer) [inline src/DynamoCoreWpf/ViewModels/Watch3D/AttachedProperties.cs]: Again, needs better docs so other developers can understand what this code is for.

File: src/DynamoCoreWpf/DynamoCoreWpf.csproj [hunk 3]
```diff
@@ -36,12 +40,24 @@
     <Prefer32Bit>false</Prefer32Bit>
   </PropertyGroup>
   <ItemGroup>
+    <Reference Include="Cyotek.Drawing.BitmapFont, Version=1.0.2.0, Culture=neutral, PublicKeyToken=58daa28b0b2de221, processorArchitecture=MSIL">
```
Comment (pboyer) [inline src/DynamoCoreWpf/DynamoCoreWpf.csproj]: What is Cyotek? Have you added a license for this?

File: extern/ProtoGeometry/ProtoGeometry.config [hunk 4]
```diff
@@ -1,4 +1,4 @@
 <?xml version="1.0"?>
 <ProtoGeometryConfiguration xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
-  <GeometryFactoryFileName>libg_219/LibG.ProtoInterface.dll</GeometryFactoryFileName>
+  <GeometryFactoryFileName>libg_220/LibG.ProtoInterface.dll</GeometryFactoryFileName>
```
Comment (pboyer) [inline extern/ProtoGeometry/ProtoGeometry.config]: It's not obvious to me why we're removing libg 219 support in this PR. Is it possible for this to be a separate PR, review by, for example, @Benglin?

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs [hunk 5]
```diff
@@ -1011,30 +1010,28 @@ private void WindowClosed(object sender, EventArgs e)
...
-            if (dynamoViewModel.BackgroundPreviewViewModel.NavigationKeyIsDown)
-            {
-                dynamoViewModel.BackgroundPreviewViewModel.NavigationKeyIsDown = false;
-                dynamoViewModel.EscapeCommand.Execute(null);
-            }
+            //if (dynamoViewModel.BackgroundPreviewViewModel.NavigationKeyIsDown)
+            //{
+            //    dynamoViewModel.BackgroundPreviewViewModel.NavigationKeyIsDown = false;
+            //    dynamoViewModel.EscapeCommand.Execute(null);
+            //}
```
Comment (pboyer) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs]: What is this? Can it be a separate PR?

File: src/DynamoCoreWpf/DynamoCoreWpf.csproj (hunk 3 above)
Comment (RodRecker) [inline src/DynamoCoreWpf/DynamoCoreWpf.csproj]: Thanks for catching this one - please add them to the 3rd party tracking sheet. See me for the link.

File: extern/ProtoGeometry/ProtoGeometry.config (hunk 4 above)
Comment (ikeough) [inline extern/ProtoGeometry/ProtoGeometry.config]: @pboyer Yeah. Sorry. That is in there because I made the change on this branch, but actually issued a separate PR from master (for you to review). I can remove those commits here.

**Comment by ikeough** [comment]
@pboyer Documentation has been added for the new types. The commits regarding MAGN-8238 and the libg_219 have been reverted. The libg_219 removal is already covered in https://github.com/DynamoDS/Dynamo/pull/5320. The fix for MAGN-8238 will be included in a separate PR.

File: src/DynamoCoreWpf/DynamoCoreWpf.csproj (hunk 3 above)
Comment (ikeough) [inline src/DynamoCoreWpf/DynamoCoreWpf.csproj]: Licenses have been added to the licenses file.

**Comment by ikeough** [comment]
Still working on the AppVeyor issue. There are configuration issues due to nuget package restore.

File: extern/ProtoGeometry/ProtoGeometry.config [hunk 6]
```diff
@@ -1,4 +1,4 @@
 <?xml version="1.0"?>
 <ProtoGeometryConfiguration xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
-  <GeometryFactoryFileName>libg_220/LibG.ProtoInterface.dll</GeometryFactoryFileName>
+  <GeometryFactoryFileName>libg_219/LibG.ProtoInterface.dll</GeometryFactoryFileName>
```
Comment (pboyer) [inline extern/ProtoGeometry/ProtoGeometry.config]: It looks like this adds all of the LibG ASM 219 files back. That sounds like a mistake...

File: src/DynamoCore/Models/NodeModelExtensions.cs [hunk 7]
```diff
@@ -47,6 +47,11 @@ internal static IEnumerable<NodeModel> UpstreamNodesMatchingPredicate(this NodeM
 
         internal static IEnumerable<NodeModel> AllUpstreamNodes(this NodeModel node, List<NodeModel> gathered)
         {
+            if(node == null)
+            {
+                return new List<NodeModel>();
+            }
```
Comment (pboyer) [inline src/DynamoCore/Models/NodeModelExtensions.cs]: I'm guessing this fixes a defect somewhere?

File: src/DynamoCoreWpf/ViewModels/Watch3D/DynamoEffectsManager.cs [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    {
+        /// <summary>
+        /// The DynamoEffectsManager is loads Effects
+        /// from shader byte code, and defines data layouts for rendering. 
+        /// By extending the DefaultEffectsManager, the DynamoEffectsManager 
+        /// makes available effects like Blinn rendering, and adds custom
+        /// Dynamo rendering effects for points, lines, and meshes.
+        /// For more information on DirectX Effects, see 
+        /// https://msdn.microsoft.com/en-us/library/windows/desktop/ff476136(v=vs.85).aspx
+        /// </summary>
```
Comment (pboyer) [inline src/DynamoCoreWpf/ViewModels/Watch3D/DynamoEffectsManager.cs]: I think this summary should be on the class, not the constructor. Right?

File: src/DynamoCoreWpf/ViewModels/Watch3D/DynamoEffectsManager.cs [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        {
+            Dispose();
+        }
+
+        /// <summary>
+        /// Override the DefaultEffectsManager's InitEffects method to load
+        /// effects used by Dynamo for rendering. Custom input layouts are
+        /// created which specify extra COLOR components used to hold data 
+        /// about the selection state and vertex coloration of objects.
+        /// See DynamoMeshVertex, DynamoPointVertex, and DynamoLineVertex
```
Comment (pboyer) [inline src/DynamoCoreWpf/ViewModels/Watch3D/DynamoEffectsManager.cs]: Consider using the cref attribute for these types:

File: src/DynamoCoreWpf/ViewModels/Watch3D/DynamoPointGeometryModel3D.cs [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+            effectTransforms = new EffectTransformVariables(effect);
+
+            var geometry = Geometry as PointGeometry3D;
+            if (geometry != null)
+            {          
+                vertexBuffer = Device.CreateBuffer(BindFlags.VertexBuffer, VertexSizeInBytes, CreateVertexArray());
+            }
+
+            vViewport = effect.GetVariableByName("vViewport").AsVector();
+            //this.vFrustum = effect.GetVariableByName("vFrustum").AsVector();
```
Comment (pboyer) [inline src/DynamoCoreWpf/ViewModels/Watch3D/DynamoPointGeometryModel3D.cs]: You know what I'm going to say about commented out code. I'd assume you have a reason to keep it - if so, document accordingly.

File: src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs [hunk 11]
```diff
@@ -321,14 +317,20 @@ public bool CanNavigateBackground
...
                 RaisePropertyChanged("CanNavigateBackground");
             }
         }
-        
+
+        public IEffectsManager EffectsManager { get; private set; }
+
+        public IRenderTechniquesManager RenderTechniquesManager { get; private set; }
+
+        public bool SupportDeferredRender { get; private set; }
```
Comment (pboyer) [inline src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs]: Add documentation for new public properties as described in the coding standards.

File: src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs [hunk 12]
```diff
@@ -682,6 +687,17 @@ internal void ComputeFrameUpdate()
             {
                 directionalLight.Direction = v;
             }
+
+            // Occasional update of scene items, including
+            // distance-based transparency sorting.
```
Comment (pboyer) [inline src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs]: Call me dumb, but I don't think this very descriptive.

File: test/Libraries/TestServices/TestSessionConfiguration.cs [hunk 13]
```diff
@@ -64,7 +64,7 @@ public TestSessionConfiguration(string dynamoCoreDirectory, string configFileDir
 
             LibraryVersion version;
             RequestedLibraryVersion = LibraryVersion.TryParse(versionStr, out version) ? 
-                version : LibraryVersion.Version220;
+                version : LibraryVersion.Version219;
```
Comment (pboyer) [inline test/Libraries/TestServices/TestSessionConfiguration.cs]: I'm guessing both of these are a mistake?

**Comment by pboyer** [comment]
@ikeough Most of my comments here are again, documentation related. All minor stuff. The code looks reasonable.

I think there's something going on here with ASM 219, so that should likely be fixed.

Is there a reason we're including all of the shader code in DynamoCoreWpf? Could it be in the extern folder? It just feels a bit funny as it's all written by someone outside of our team. Maybe you anticipate it will be edited by us later?

Anyhow, this LGTM once those are fixed. :)

**Comment by pboyer** [comment]
@aparajit-pratap Can you say a bit about the EngOps Build? It feels like it erroneously fails a lot, but I'm not sure. Can we keep this on a fork until it's accurate?

File: test/Libraries/TestServices/TestSessionConfiguration.cs (hunk 13 above)
Comment (ikeough) [inline test/Libraries/TestServices/TestSessionConfiguration.cs]: Nope. Not a mistake. This was the reversion of the original commits to remove libg219 support. That was applied on master, and also on this branch. Then we reverted it on this branch which makes it seem like I'm making this change again. When we ultimately merge this back to master, I'll get a merge conflict and have to resolve it then.

File: src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs (hunk 12 above)
Comment (ikeough) [inline src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs]: Ok, "dumb, but I don't think this is very descriptive." Are you sure you want me to call you that all the time? It's pretty long. I'll probably make a mistake every once and a while.

**Comment by ikeough** [comment]
The build is failing because the EngOps machines use VS2013, which doesn't have package restore on by default, like VS2015. I'm fixing today.
